src/shims/runtime/logger.js

```diff
@@ -0,0 +1,103 @@
+/* eslint-disable no-console */
+'use strict'
+
+// Loosely based on Winston's Logger logic,
+// simplified for our minimal needs.
+// See license at https://github.com/winstonjs/winston/blob/master/LICENSE
+// for original source code.
+
+function getLogger () {
+  return {
+    /**
+     * @param {any[]} args
+     */
+    debug (...args) {
+      return this.log(console.debug, ...args)
+    },
+
+    /**
+     * @param {any[]} args
+     */
+    info (...args) {
+      return this.log(console.info, ...args)
+    },
+
+    /**
+     * @param {any[]} args
+     */
+    warn (...args) {
+      return this.log(console.warn, ...args)
+    },
+
+    /**
+     * @param {any[]} args
+     */
+    error (...args) {
+      return this.log(console.error, ...args)
+    },
+
+    /**
+     * @param {CallableFunction} fn
+     * @param {any[]} args
+     */
+    log (fn, ...args) {
+      // Optimize the hot-path which is the single object.
+      if (args.length === 1) {
+        const [msg] = args
+        const info = (msg && msg.message && msg) || { message: msg }
+        this._write(fn, info)
+        return this
+      }
+
+      // When provided nothing assume the empty string
+      if (args.length === 0) {
+        this._log(fn, '')
+        return this
+      }
+
+      // Otherwise build argument list
+      return this._log(fn, args[0], ...args.slice(1))
+    },
+
+    /**
+     * @param {CallableFunction} fn
+     * @param {any} msg
+     * @param {any[]} splat
+     */
+    _log (fn, msg, ...splat) {
+      // eslint-disable-line max-params
+      // Optimize for the hotpath of logging JSON literals
+      if (arguments.length === 1) {
+        this._write(fn, '')
+        return this
+      }
+
+      // Slightly less hotpath, but worth optimizing for.
+      if (arguments.length === 2) {
+        if (msg && typeof msg === 'object') {
+          this._write(fn, msg)
+          return this
+        }
+
+        this._write(fn, { message: msg })
+        return this
+      }
+
+      this._write(fn, { message: msg, splat })
+      return this
+    },
+
+    /**
+     * @param {CallableFunction} fn
+     * @param {any} msg
+     */
+    _write (fn, msg) {
+      const s = JSON.stringify(msg)
+      fn(s)
```

Comment:
I thought we'd just return `console`, why are we using this code? It's quite complex, with optimization, stringify and custom log functions.

Reply:
Because the way we actually log from our codebase, using just `console` wouldn't give you any info at all, as it doesn't handle the given arguments well. Like that you just see things like `logged IMessage (IEvent)` or w/e, without actually seeing what's inside.

I agree that it's not as simple as using `console`, but neither do I think it's "that complex". I think this is the price you pay or not adding that support directly to winston as a contribution. That or not using winston at all, but that will result in even more code by using a complete solution.

Reply:
Thanks for the details, I understand why we need to move around JSON objects.

I still find the code is more complex than it should be, we're already optimizing 4 different cases, with 6 different branches. If you think that's really necessary, could you add a test for each of these cases?

it looks like we need at least:

```
logger.warn()
logger.warn('helloworld')
logger.warn({some: 'object', with: 'values'})
logger.warn('message', 'additional', 'parameters')
logger.warn('message', {additional: 'parameters'))
```

Reply:
Resolved in 36fca7f6cac6e5407c07bd73059f51c566dc3e1b.